superduperdb/backends/base/query.py

```diff
@@ -1,761 +1,268 @@
-        if self.query_linker is not None:
-            query_linker = getattr(self.query_linker, name)
+    def unique_id(self):
+        return self.id
+
+    def _deep_flat_encode(self, cache, blobs, files, leaves_to_keep=()):
+        query, documents = self._dump_query()
+        cache[self.id] = {
+            '_path': 'superduperdb.backends.base.new_query.parse_query',
+            'documents': documents,
+            'query': query,
+        }
+        return self.id
+
+    @staticmethod
+    def _update_item(a, documents, queries):
+        if isinstance(a, Query):
+            a, sub_documents, sub_queries = a._to_str()
+            documents.update(sub_documents)
+            queries.update(sub_queries)
+            id_ = uuid.uuid4().hex[:5].upper()
+            queries[id_] = a
+            arg = f'query[{id_}]'
         else:
-            query_linker = self._get_query_linker(
-                self.table_or_collection,
-                members=[self._get_query_component(name, type=QueryType.ATTR)],
+            id_ = uuid.uuid4().hex[:5].upper()
+            if isinstance(a, dict):
+                documents[id_] = a
+                arg = f'documents[{id_}]'
+            elif isinstance(a, list):
+                documents[id_] = {'_base': a}
+                arg = f'documents[{id_}]'
+            else:
+                try:
+                    arg = json.dumps(a)
+                except Exception:
+                    documents[id_] = {'_base': a}
+                    arg = id_
+        return arg
+
+    def _to_str(self):
+        documents = {}
+        queries = {}
+        out = self.identifier[:]
+        for part in self.parts:
+            if isinstance(part, str):
+                out += f'.{part}'
+                continue
+            args = []
+            for a in part[1]:
+                args.append(self._update_item(a, documents, queries))
+            args = ', '.join(args)
+            kwargs = {}
+            for k, v in kwargs.items():
+                kwargs[k] = self._update_item(v, documents, queries)
+            kwargs = ', '.join([f'{k}={v}' for k, v in kwargs.items()])
+            if part[1] and part[2]:
+                out += f'.{part[0]}({args}, {kwargs})'
+            if not part[1] and part[2]:
+                out += f'.{part[0]}({kwargs})'
+            if part[1] and not part[2]:
+                out += f'.{part[0]}({args})'
+            if not part[1] and not part[2]:
+                out += f'.{part[0]}()'
+        return out, documents, queries
+
+    def _dump_query(self):
+        output, documents, queries = self._to_str()
+        if queries:
+            output = (
+                '\n'.join(list(queries.values()))
+                + '\n' + output
             )
-        return self._query_from_parts(
-            table_or_collection=self.table_or_collection,
-            pre_like=self.pre_like,
-            query_linker=query_linker,
+        for i, k in enumerate(queries):
```

Comment:
I see we are replacing random ids with sequence in documents dictionary...
This makes the query serialisation shorter!